File: db4oj.tests/src/com/db4o/db4ounit/common/foundation/Collection4TestCase.java
/* Copyright (C) 2004 - 2006  db4objects Inc.  http://www.db4o.com */

package com.db4o.db4ounit.common.foundation;

import com.db4o.foundation.*;

import db4ounit.*;

public class Collection4TestCase implements TestCase {
	
	public static void main(String[] args) {
		new TestRunner(Collection4TestCase.class).run();
	}
	
	public void testRemoveAll() {
		final String[] originalElements = new String[] { "foo", "bar", "baz" };
		final Collection4 c = newCollection(originalElements);
		c.removeAll(newCollection(new String[0]));
		assertCollection(originalElements, c);
		
		c.removeAll(newCollection(new String[] { "baz", "bar", "zeng" }));
		assertCollection(new String[] { "foo" }, c);
		
		c.removeAll(newCollection(originalElements));
		assertCollection(new String[0], c);
		 
	}
	
	public void testReplace(){
        final Collection4 c = new Collection4();
        c.replace("one", "two");
        c.add("one");
        c.add("two");
        c.add("three");
        c.replace("two", "two.half");
        assertCollection(new String[] {"one", "two.half", "three"}, c);
        c.replace("two.half", "one");
        c.replace("one", "half");
        assertCollection(new String[] {"half", "one", "three"}, c);
	}
	
	public void testNulls(){
		final Collection4 c = new Collection4();
		c.add("one");
		assertNotContainsNull(c);
		c.add(null);
		assertContainsNull(c);
		assertCollection(new String[] { "one", null }, c);
		c.prepend(null);
		assertCollection(new String[] { null, "one", null }, c);
		c.prepend("zero");
		c.add("two");
		assertCollection(new String[] {"zero", null, "one", null , "two"}, c);
		assertContainsNull(c);
		c.remove(null);
		assertCollection(new String[] { "zero", "one", null , "two"}, c);
		c.remove(null);
		assertNotContainsNull(c);
		assertCollection(new String[] { "zero", "one", "two"}, c);
		c.remove(null);
		assertCollection(new String[] { "zero", "one", "two"}, c);
	}
	
	public void testPrepend() {
		final Collection4 c = new Collection4();
		c.prepend("foo");
		assertCollection(new String[] { "foo" }, c);
		c.add("bar");
		assertCollection(new String[] { "foo", "bar" }, c);
		c.prepend("baz");
		assertCollection(new String[] { "baz", "foo", "bar" }, c);
		c.prepend("gazonk");
		assertCollection(new String[] { "gazonk", "baz", "foo", "bar" }, c);
	}
	
	public void testCopyConstructor() {
		final String[] expected = new String[] { "1", "2", "3" };
		final Collection4 c = newCollection(expected);
		assertCollection(expected, new Collection4(c));
	}
	
	public void testInvalidIteratorException() {
		final Collection4 c = newCollection(new String[] { "1", "2" });
		final Iterator4 i = c.iterator();
		Assert.isTrue(i.moveNext());
		c.add("3");
		Assert.expect(InvalidIteratorException.class, new CodeBlock() {
			public void run() throws Throwable {
				System.out.println(i.current());
			}
		});
	}
	
	public void testRemove() {
		final Collection4 c = newCollection(new String[] { "1", "2", "3", "4" });
		c.remove("3");
		assertCollection(new String[] { "1", "2", "4"} , c);
		c.remove("4");
		assertCollection(new String[] { "1", "2" } , c);
		c.add("5");
		assertCollection(new String[] { "1", "2", "5" } , c);
		c.remove("1");
		assertCollection(new String[] { "2", "5" } , c);
		c.remove("2");
		c.remove("5");
		assertCollection(new String[] {}, c);
		c.add("6");
		assertCollection(new String[] { "6" }, c);
	}
	
	private void assertCollection(String[] expected, Collection4 c) {
		Assert.areEqual(expected.length, c.size());
		Iterator4Assert.areEqual(expected, c.iterator());
	}
	
	private void assertContainsNull(Collection4 c) {
		Assert.isTrue(c.contains(null));
		Assert.isNull(c.get(null));
		int size = c.size();
		c.ensure(null);
		Assert.areEqual(size, c.size());
	}
	
	private void assertNotContainsNull(Collection4 c) {
		Assert.isFalse(c.contains(null));
		Assert.isNull(c.get(null));
		int size = c.size();
		c.ensure(null);
		Assert.areEqual(size + 1, c.size());
		c.remove(null);
		Assert.areEqual(size, c.size());
	}

	public void testIterator() {
		String[] expected = new String[] { "1", "2", "3" };
		Collection4 c = newCollection(expected);		
		Iterator4Assert.areEqual(expected, c.iterator());
	}	
	
	private Collection4 newCollection(String[] expected) {
		Collection4 c = new Collection4();		
		c.addAll(expected);
		return c;
	}
	
	public void testToString() {
		Collection4 c = new Collection4();
		Assert.areEqual("[]", c.toString());
		
		c.add("foo");
		Assert.areEqual("[foo]", c.toString());
		c.add("bar");
		Assert.areEqual("[foo, bar]", c.toString());
	}
}

File: db4on/test/src/com/db4o/db4ounit/common/foundation/Collection4TestCase.cs
namespace com.db4o.db4ounit.common.foundation
{
	public class Collection4TestCase : Db4oUnit.TestCase
	{
		public static void Main(string[] args)
		{
			new Db4oUnit.TestRunner(typeof(com.db4o.db4ounit.common.foundation.Collection4TestCase)
				).Run();
		}

		public virtual void TestNulls()
		{
			com.db4o.foundation.Collection4 c = new com.db4o.foundation.Collection4();
			c.Add("one");
			AssertNotContainsNull(c);
			c.Add(null);
			AssertContainsNull(c);
			AssertCollection(new string[] { "one", null }, c);
			c.Prepend(null);
			AssertCollection(new string[] { null, "one", null }, c);
			c.Prepend("zero");
			c.Add("two");
			AssertCollection(new string[] { "zero", null, "one", null, "two" }, c);
			AssertContainsNull(c);
			c.Remove(null);
			AssertCollection(new string[] { "zero", "one", null, "two" }, c);
			c.Remove(null);
			AssertNotContainsNull(c);
			AssertCollection(new string[] { "zero", "one", "two" }, c);
			c.Remove(null);
			AssertCollection(new string[] { "zero", "one", "two" }, c);
		}

		public virtual void TestPrepend()
		{
			com.db4o.foundation.Collection4 c = new com.db4o.foundation.Collection4();
			c.Prepend("foo");
			AssertCollection(new string[] { "foo" }, c);
			c.Add("bar");
			AssertCollection(new string[] { "foo", "bar" }, c);
			c.Prepend("baz");
			AssertCollection(new string[] { "baz", "foo", "bar" }, c);
			c.Prepend("gazonk");
			AssertCollection(new string[] { "gazonk", "baz", "foo", "bar" }, c);
		}

		public virtual void TestCopyConstructor()
		{
			string[] expected = new string[] { "1", "2", "3" };
			com.db4o.foundation.Collection4 c = NewCollection(expected);
			AssertCollection(expected, new com.db4o.foundation.Collection4(c));
		}

		public virtual void TestInvalidIteratorException()
		{
			com.db4o.foundation.Collection4 c = NewCollection(new string[] { "1", "2" });
			System.Collections.IEnumerator i = c.GetEnumerator();
			Db4oUnit.Assert.IsTrue(i.MoveNext());
			c.Add("3");
			Db4oUnit.Assert.Expect(typeof(com.db4o.foundation.InvalidIteratorException), new 
				_AnonymousInnerClass60(this, i));
		}

		private sealed class _AnonymousInnerClass60 : Db4oUnit.CodeBlock
		{
			public _AnonymousInnerClass60(Collection4TestCase _enclosing, System.Collections.IEnumerator
				 i)
			{
				this._enclosing = _enclosing;
				this.i = i;
			}

			public void Run()
			{
				j4o.lang.JavaSystem.Out.WriteLine(i.Current);
			}

			private readonly Collection4TestCase _enclosing;

			private readonly System.Collections.IEnumerator i;
		}

		public virtual void TestRemove()
		{
			com.db4o.foundation.Collection4 c = NewCollection(new string[] { "1", "2", "3", "4"
				 });
			c.Remove("3");
			AssertCollection(new string[] { "1", "2", "4" }, c);
			c.Remove("4");
			AssertCollection(new string[] { "1", "2" }, c);
			c.Add("5");
			AssertCollection(new string[] { "1", "2", "5" }, c);
			c.Remove("1");
			AssertCollection(new string[] { "2", "5" }, c);
			c.Remove("2");
			c.Remove("5");
			AssertCollection(new string[] {  }, c);
			c.Add("6");
			AssertCollection(new string[] { "6" }, c);
		}

		private void AssertCollection(string[] expected, com.db4o.foundation.Collection4 
			c)
		{
			Db4oUnit.Assert.AreEqual(expected.Length, c.Size());
			AssertIterator(expected, c.GetEnumerator());
		}

		private void AssertContainsNull(com.db4o.foundation.Collection4 c)
		{
			Db4oUnit.Assert.IsTrue(c.Contains(null));
			Db4oUnit.Assert.IsNull(c.Get(null));
			int size = c.Size();
			c.Ensure(null);
			Db4oUnit.Assert.AreEqual(size, c.Size());
		}

		private void AssertNotContainsNull(com.db4o.foundation.Collection4 c)
		{
			Db4oUnit.Assert.IsFalse(c.Contains(null));
			Db4oUnit.Assert.IsNull(c.Get(null));
			int size = c.Size();
			c.Ensure(null);
			Db4oUnit.Assert.AreEqual(size + 1, c.Size());
			c.Remove(null);
			Db4oUnit.Assert.AreEqual(size, c.Size());
		}

		public virtual void TestIterator()
		{
			string[] expected = new string[] { "1", "2", "3" };
			com.db4o.foundation.Collection4 c = NewCollection(expected);
			AssertIterator(expected, c.GetEnumerator());
		}

		private com.db4o.foundation.Collection4 NewCollection(string[] expected)
		{
			com.db4o.foundation.Collection4 c = new com.db4o.foundation.Collection4();
			c.AddAll(expected);
			return c;
		}

		private void AssertIterator(string[] expected, System.Collections.IEnumerator iterator
			)
		{
			Db4oUnit.Assert.IsNotNull(iterator);
			for (int i = 0; i < expected.Length; ++i)
			{
				Db4oUnit.Assert.IsTrue(iterator.MoveNext());
				Db4oUnit.Assert.AreEqual(expected[i], iterator.Current);
			}
			Db4oUnit.Assert.IsFalse(iterator.MoveNext());
		}

		public virtual void TestToString()
		{
			com.db4o.foundation.Collection4 c = new com.db4o.foundation.Collection4();
			Db4oUnit.Assert.AreEqual("[]", c.ToString());
			c.Add("foo");
			Db4oUnit.Assert.AreEqual("[foo]", c.ToString());
			c.Add("bar");
			Db4oUnit.Assert.AreEqual("[foo, bar]", c.ToString());
		}
	}
}
